### PR TITLE
Block Library: Fix Query block undo trap during creation

### LINKED ISSUE #29494: Query Block: allow for the ability to undo setting changes
## Description

As part of the [second FSE call for testing](https://make.wordpress.org/test/2021/02/18/fse-program-testing-call-2-build-a-homepage-with-site-editing-blocks/#comment-1362), a person who responded with feedback found that they were unable to undo changes to the settings when using the Query block after initially selecting how the block should display.

## Step-by-step reproduction instructions

1. Use a block based theme and the latest version of Gutenberg to setup the FSE experience.
2. Go to "Site Editor (beta)"
3. Add a query block
4. Select an initial setup (Post title with excerpt, etc)
5. Try to undo the selection. Find out that you can't!

## Expected behaviour

I expected to be able to undo setting changes. 

## Actual behaviour

I was unable to undo changes to the initial query block setup. 

## Screenshots or screen recording (optional)

https://user-images.githubusercontent.com/26996883/109733591-a3406680-7b7c-11eb-96ce-8bd7bb6dcb9a.mov

## WordPress information
- WordPress version: 5.6.2
- Gutenberg version: 10.0.2
- Are all plugins except Gutenberg deactivated? Yes
- Are you using a default theme (e.g. Twenty Twenty-One)? TT1 Blocks

### PR DESCRIPTION
**Pull request by ntsekouras**
## Description
Fixes: https://github.com/WordPress/gutenberg/issues/29494

In `Query` block there are some effects running where some initialization logic is happening and setting some values to some attributes (ex. queryId).

These updates can cause an `undo trap` where undoing will result in resetting again, so we need to mark these changes as not persistent with `__unstableMarkNextChangeAsNotPersistent`.

## Testing instructions
1. Insert a Query block by selecting a pattern or a block variation
2. Undo
3. Observe that the `undo` is performed properly and everything else works as before and expected.

## Screenshots 
queryundo

### CODE REVIEW

File: packages/block-library/src/query/edit/index.js [hunk 1]
```diff
@@ -41,13 +50,15 @@ export function QueryContent( { attributes, setAttributes } ) {
 			newQuery.perPage = postsPerPage;
 		}
 		if ( !! Object.keys( newQuery ).length ) {
+			__unstableMarkNextChangeAsNotPersistent();
 			updateQuery( newQuery );
 		}
-	}, [ query.perPage, query.inherit ] );
+	}, [ query.perPage ] );
```
Comment (ntsekouras) [inline packages/block-library/src/query/edit/index.js:56]: This is unrelated and just a leftover..

**Review by Addison-Stavlo — APPROVED** [review]
Tested, and this fixes the undo trap as expected!  Undo/redo and save flows seem to work as expected.

Unrelated to the change here, I noticed some errors in the console related to the post title block:
Screen Shot 2021-03-24 at 2 21 04 PM

and opened a quick fix PR https://github.com/WordPress/gutenberg/pull/30208

**Comment by Addison-Stavlo** [comment]
🤔  Noticing a small thing, although cannot consistently reproduce:

If i insert query block -> select a pattern -> save the changes -> select 'undo': it seems "redo" is sometimes not an available option.  It seems small and overall this feels like a big step up from where it was, but noticing that inconsistency at times.

**Comment by ellatrix** [comment]
Oh, do you think this could work for the image block too? #30114

**Review by annezazu — COMMENTED** [review]
Can't comment on the code but the experience worked great based on my testing! Thanks for getting this in place - it's come up a few times with the FSE outreach program:

https://user-images.githubusercontent.com/26996883/112383630-ff496700-8cb2-11eb-85e0-461f52102d0f.mov

The only thing I noticed is that the site content was all out of wack in terms of width but I'm guessing that's due to the recent PR exploring changing how alignments are handled with FSE & block themes.

**Review by carolinan — APPROVED** [review]
I was able to replicate the problem, and in my tests this PR solves the issue.
I am seeing the same JS error mentioned above, but the redo is working well for me.

**Comment by ntsekouras** [comment]
I noticed this error yesterday and meant to see the root cause of this tomorrow.

In my quick check I think it has to do with `isLink` handling in Post Title, but I'll look at it tomorrow if it's not resolved by then :)

>Oh, do you think this could work for the image block too? #30114

I guess so, it seems like a similar problem. I'll check the PR better tomorrow though and it's worth trying I think @ellatrix .

**Pull request merged**